pmd-java/src/test/resources/net/sourceforge/pmd/lang/java/metrics/impl/xml/AtfdTest.xml

```diff
@@ -0,0 +1,188 @@
+        dataFlow.createNewNode(node);
+        return super.visit(node, data);
+    }
+
+    @Override
+    public Object visit(ASTExpression node, Object data) {
+        if (!(data instanceof Structure)) {
+            return data;
+        }
+        Structure dataFlow = (Structure) data;
+
+        String loggerTag = "parent";
+
+        Node parent = node.jjtGetParent();
+
+        // TODO what about throw stmts?
+        if (parent instanceof ASTIfStatement) {
+            dataFlow.createNewNode(node); // START IF
+            dataFlow.pushOnStack(NodeType.IF_EXPR, dataFlow.getLast());
+            tryToLog(loggerTag, NodeType.IF_EXPR, node);
+        } else if (parent instanceof ASTWhileStatement) {
+            dataFlow.createNewNode(node); // START WHILE
+            dataFlow.pushOnStack(NodeType.WHILE_EXPR, dataFlow.getLast());
+            tryToLog(loggerTag, NodeType.WHILE_EXPR, node);
+        } else if (parent instanceof ASTSwitchStatement) {
+            dataFlow.createNewNode(node); // START SWITCH
+            dataFlow.pushOnStack(NodeType.SWITCH_START, dataFlow.getLast());
+            tryToLog(loggerTag, NodeType.SWITCH_START, node);
+        } else if (parent instanceof ASTForStatement) {
+            dataFlow.createNewNode(node); // FOR EXPR
+            dataFlow.pushOnStack(NodeType.FOR_EXPR, dataFlow.getLast());
+            tryToLog(loggerTag, NodeType.FOR_EXPR, node);
+        } else if (parent instanceof ASTDoStatement) {
+            dataFlow.createNewNode(node); // DO EXPR
+            dataFlow.pushOnStack(NodeType.DO_EXPR, dataFlow.getLast());
+            tryToLog(loggerTag, NodeType.DO_EXPR, node);
+        } else if (parent instanceof ASTAssertStatement) {
+            dataFlow.createNewNode(node);
+            dataFlow.pushOnStack(NodeType.ASSERT_STATEMENT, dataFlow.getLast());
+            tryToLog(loggerTag, NodeType.ASSERT_STATEMENT, node);
+        }
+
+        return super.visit(node, data);
+    }
+
+}
+
+     ]]></code-fragment>
+
+
+    <test-code>
+        <description>Full example</description> <!-- TODO issues w/ that-->
+        <expected-problems>7</expected-problems>
+        <expected-messages>
+            <message>'.StatementAndBraceFinder' has value 10 highest 6.</message>
+            <message>'.StatementAndBraceFinder#buildDataFlowFor(JavaNode)' has value 6.</message>
+            <message>'.StatementAndBraceFinder#tryToLog(String, NodeType, Node)' has value 4.</message>
+            <message>'.StatementAndBraceFinder#tryToLog(NodeType, Node)' has value 0.</message>
+            <message>'.StatementAndBraceFinder#visit(ASTStatementExpression, Object)' has value 4.</message>
+            <message>'.StatementAndBraceFinder#visit(ASTVariableDeclarator, Object)' has value 4.</message>
+            <message>'.StatementAndBraceFinder#visit(ASTExpression, Object)' has value 18.</message>
+        </expected-messages>
+        <code-ref id="full-example"/>
+    </test-code>
+
+    <test-code>
```

Comment:
you can add the attribute `regressionTest="true`", then this test case will be ignored.... You can then expect the correct message.